13 change date to unavailable

russellvaughan opened the pull request:
closes #13

russellvaughan committed:
fixed request.space.id bug on requests.erb

russellvaughan committed:
WIP user cannot confirm dates if already booked

russellvaughan committed:
User story 13 complete. Created a booking table. included unavailable dates in viewe_space.erb. updated view to display unavailable dates. prevented from booking already booked dates / or dates that aren't in range. Add helper method to app.rb to validate request datesUser story 13 complete. Created a booking table. included unavailable dates in viewe_space.erb. updated view to display unavailable dates. prevented from booking already booked dates / or dates that aren't in range. Add helper method to app.rb to validate request dates..

jaxdid commented on app/app.rb:
```diff
@@ -20,16 +20,34 @@ def current_user
+      end
+    end
+
+    def check_booking_to_date(date, range)
```
Is `#check_booking_to_date` functionally identical to `#check_booking_from_date`? Can they be consolidated into a single method (e.g., `#check_booking_validity`)?

jaxdid commented on app/controllers/requests.rb:
```diff
@@ -1,15 +1,21 @@
+    booking_to = space.bookings.map(&:end_date)
+    request_range=(params[:start_date]..params[:end_date])
+    check_booking_from_date(booking_from, request_range)
+    check_booking_to_date(booking_to, request_range)
```
If consolidating `#check_booking_from` and `#check_booking_to` (and refactoring the resulting method), would it be logical to do a single method call, passing in the three arguments (e.g., `check_booking_validity(booking_from, booking_to, request_range)`)?

jaxdid commented on app/models/booking.rb:
```diff
@@ -0,0 +1,11 @@
+class Booking 
```
Indent lines 2-10 :)

russellvaughan committed:
Built in edge cases for user submitting dates before or after availability.
refactored 'check_booking for/to' methods into one 'check_booking' method

jaxdid commented on app/views/layout.erb:
```diff
@@ -24,6 +24,15 @@
 
         </div>
     </header>
+    <% if flash[:booked] && !flash[:booked].empty? %>
```
Is a distinct `<ul id="booked">` necessary? I guess because we would not want to style the message the same as, say, invalid form data alerts (in `<ul id="errors">`)?

tcpickard94 replied:
We thought it was for the reasons you said above. Not 100% sure we need the whole each loop though, we could do it like the saved one below, any thoughts?

jaxdid commented on app/views/view_space.erb:
```diff
@@ -2,7 +2,14 @@
     <div>
-        Available From: <%= format_date(@space.date_from) %> Available To: <%= format_date(@space.date_to) %>
+
+    Available From: <%= format_date(@space.date_from) %> Available To: <%= format_date(@space.date_to) %>
```
Indentation, lines 6-10

jaxdid commented on app/app.rb:
```diff
@@ -20,16 +20,34 @@ def current_user
 
+    def check_booking_from_date(date, range)
+      date.each do |booking|
+        if range.include?(booking.to_s) 
```
Check that booking requests for dates entirely outside of (before or after) a space's availability are rejected/not entered into the database.

tcpickard94 committed:
Refactored all files for consistent indentation and spacing

tcpickard94 committed:
Refactored User Story 13

Moved some validation logic from the request controller to several
helpers in app

jaxdid commented:
:shipit:
